File: app/src/main/java/com/gproust/sprout/ui/health/HealthScreen.kt
package com.gproust.sprout.ui.health

import android.content.Context
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.compose.viewModel
import com.gproust.sprout.R
import com.gproust.sprout.data.SproutRepository
import com.gproust.sprout.data.local.Bleeding
import com.gproust.sprout.data.local.BreastState
import com.gproust.sprout.data.local.DeliveryType
import com.gproust.sprout.data.local.Recovery
import com.gproust.sprout.data.local.WellbeingEntity
import com.gproust.sprout.ui.common.ChoiceChips
import com.gproust.sprout.ui.common.EmptyHint
import com.gproust.sprout.ui.common.EntryCard
import com.gproust.sprout.ui.common.FieldLabel
import com.gproust.sprout.ui.common.NotesField
import com.gproust.sprout.ui.common.SproutTopBar
import com.gproust.sprout.ui.common.TimePickerField
import com.gproust.sprout.ui.common.formatDateTime
import com.gproust.sprout.ui.common.healingFieldLabel
import com.gproust.sprout.ui.common.label
import com.gproust.sprout.ui.common.moodEmoji
import com.gproust.sprout.ui.rememberSproutViewModelFactory
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class HealthViewModel(private val repository: SproutRepository) : ViewModel() {
    val gaveBirth = repository.parentProfile
        .map { it?.gaveBirth ?: false }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), false)
    val breastfeeding = repository.parentProfile
        .map { it?.breastfeeding ?: false }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), false)
    val deliveryType = repository.parentProfile
        .map { it?.deliveryType }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), null)

    val entries = repository.wellbeing
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun add(entity: WellbeingEntity) = viewModelScope.launch { repository.addWellbeing(entity) }
    fun delete(entity: WellbeingEntity) = viewModelScope.launch { repository.deleteWellbeing(entity) }
}

@Composable
fun HealthScreen(onBack: () -> Unit) {
    val vm: HealthViewModel = viewModel(factory = rememberSproutViewModelFactory())
    val gaveBirth by vm.gaveBirth.collectAsState()
    val breastfeeding by vm.breastfeeding.collectAsState()
    val deliveryType by vm.deliveryType.collectAsState()
    val entries by vm.entries.collectAsState()
    val context = LocalContext.current

    Scaffold(topBar = { SproutTopBar(stringResource(R.string.screen_wellbeing), onBack = onBack) }) { padding ->
        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(padding),
            contentPadding = PaddingValues(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            item {
                WellbeingAddCard(
                    gaveBirth = gaveBirth,
                    breastfeeding = breastfeeding,
                    deliveryType = deliveryType,
                    onAdd = vm::add,
                )
            }
            item {
                Text(
                    stringResource(R.string.history),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                )
            }
            if (entries.isEmpty()) {
                item { EmptyHint(stringResource(R.string.wellbeing_empty)) }
            }
            items(entries, key = { it.id }) { entry ->
                EntryCard(
                    title = stringResource(R.string.wellbeing_mood_title, moodEmoji(entry.mood), entry.mood),
                    subtitle = wellbeingSubtitle(context, entry),
                    meta = formatDateTime(context, entry.time),
                    icon = Icons.Filled.Favorite,
                    onDelete = { vm.delete(entry) },
                )
            }
        }
    }
}

private fun wellbeingSubtitle(context: Context, entry: WellbeingEntity): String {
    val parts = buildList {
        entry.recovery?.let { add(context.getString(R.string.wellbeing_sub_healing, it.label(context))) }
        entry.bleeding?.let { add(context.getString(R.string.wellbeing_sub_bleeding, it.label(context))) }
        entry.breast?.let { add(context.getString(R.string.wellbeing_sub_breasts, it.label(context))) }
        if (!entry.notes.isNullOrBlank()) add(entry.notes)
    }
    return parts.joinToString(context.getString(R.string.feeding_detail_separator))
}

@Composable
private fun WellbeingAddCard(
    gaveBirth: Boolean,
    breastfeeding: Boolean,
    deliveryType: DeliveryType?,
    onAdd: (WellbeingEntity) -> Unit,
) {
    val context = LocalContext.current
    var mood by remember { mutableIntStateOf(3) }
    var bleeding by remember { mutableStateOf<Bleeding?>(null) }
    var recovery by remember { mutableStateOf<Recovery?>(null) }
    var breast by remember { mutableStateOf<BreastState?>(null) }
    var time by remember { mutableLongStateOf(System.currentTimeMillis()) }
    var notes by remember { mutableStateOf("") }

    Card {
        Column(Modifier.padding(16.dp)) {
            Text(stringResource(R.string.wellbeing_add_title), style = MaterialTheme.typography.titleMedium)

            FieldLabel(stringResource(R.string.field_mood))
            ChoiceChips(
                options = listOf(1, 2, 3, 4, 5),
                selected = mood,
                onSelect = { mood = it },
                labelOf = { "${moodEmoji(it)} $it" },
            )

            if (gaveBirth) {
                FieldLabel(healingFieldLabel(context, deliveryType))
                ChoiceChips(
                    options = Recovery.entries,
                    selected = recovery,
                    onSelect = { recovery = if (recovery == it) null else it },
                    labelOf = { it.label(context) },
                )

                FieldLabel(stringResource(R.string.field_bleeding))
                ChoiceChips(
                    options = Bleeding.entries,
                    selected = bleeding,
                    onSelect = { bleeding = if (bleeding == it) null else it },
                    labelOf = { it.label(context) },
                )
            }

            if (breastfeeding) {
                FieldLabel(stringResource(R.string.field_breast_comfort))
                ChoiceChips(
                    options = BreastState.entries,
                    selected = breast,
                    onSelect = { breast = if (breast == it) null else it },
                    labelOf = { it.label(context) },
                )
            }

            FieldLabel(stringResource(R.string.field_time))
            TimePickerField(label = stringResource(R.string.picker_at), millis = time, onChange = { time = it })

            Spacer(Modifier.height(8.dp))
            NotesField(value = notes, onChange = { notes = it })

            Spacer(Modifier.height(12.dp))
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
                Button(onClick = {
                    onAdd(
                        WellbeingEntity(
                            time = time,
                            mood = mood,
                            bleeding = if (gaveBirth) bleeding else null,
                            recovery = if (gaveBirth) recovery else null,
                            breast = if (breastfeeding) breast else null,
                            notes = notes.ifBlank { null },
                        ),
                    )
                    mood = 3
                    bleeding = null
                    recovery = null
                    breast = null
                    notes = ""
                    time = System.currentTimeMillis()
                }) {
                    Text(stringResource(R.string.wellbeing_add))
                }
            }
        }
    }
}

File: ios/Sprout/Sources/Screens/Health/HealthScreen.swift
import SproutData
import SwiftUI

/// The parent's own record, from `ui/health/HealthScreen.kt`.
///
/// Belongs to the parent rather than to a baby (BDR-0001): no `babyId`, and it
/// is the one table that never leaves the phone — not to a partner's handset,
/// not into a backup's sync payload. That is why `deleteWellbeing` is a real
/// delete and not the soft delete every other log gets: there is no merge for
/// the deletion to have to survive.
@Observable
@MainActor
final class HealthViewModel {
    var entries: [Wellbeing] = []
    var gaveBirth = false
    var breastfeeding = false
    var deliveryType: DeliveryType?

    private let repository: SproutRepository

    init(repository: SproutRepository) {
        self.repository = repository
    }

    func observeEverything() async {
        let repository = self.repository
        await withTaskGroup(of: Void.self) { group in
            group.addTask {
                await observe(repository.wellbeing) { [weak self] in self?.entries = $0 }
            }
            group.addTask {
                await observe(repository.parentProfile) { [weak self] profile in
                    self?.gaveBirth = profile?.gaveBirth ?? false
                    self?.breastfeeding = profile?.breastfeeding ?? false
                    self?.deliveryType = profile?.deliveryType
                }
            }
        }
    }

    func add(_ entry: Wellbeing) { try? repository.addWellbeing(entry) }
    func delete(_ entry: Wellbeing) { try? repository.deleteWellbeing(entry) }
}

struct HealthScreen: View {
    @Environment(\.sprout) private var sprout
    @State private var model: HealthViewModel?
    @State private var pendingDelete: Wellbeing?

    var body: some View {
        Group {
            if let model { content(model) } else { Color.clear }
        }
        .navigationTitle(Str.t("screen_wellbeing"))
        .navigationBarTitleDisplayMode(.inline)
        .task {
            let model = model ?? HealthViewModel(repository: sprout.repository)
            self.model = model
            await model.observeEverything()
        }
    }

    @ViewBuilder
    private func content(_ model: HealthViewModel) -> some View {
        ScrollView {
            LazyVStack(alignment: .leading, spacing: Spacing.snug) {
                WellbeingAddCard(
                    gaveBirth: model.gaveBirth,
                    breastfeeding: model.breastfeeding,
                    deliveryType: model.deliveryType,
                    onAdd: { model.add($0) }
                )

                SectionLabel(Str.t("history"))
                if model.entries.isEmpty {
                    EmptyHint(Str.t("wellbeing_empty"))
                }
                ForEach(model.entries) { entry in
                    EntryCard(
                        title: Str.t("wellbeing_mood_title", moodEmoji(entry.mood), entry.mood),
                        subtitle: subtitle(entry),
                        meta: SproutDateStyle.dateTime(entry.time),
                        systemImage: "heart.fill",
                        onDelete: { pendingDelete = entry }
                    )
                }
            }
            .padding(Spacing.regular)
        }
        .sproutStyle()
        .confirmDelete(
            isPresented: Binding(get: { pendingDelete != nil }, set: { if !$0 { pendingDelete = nil } })
        ) {
            if let entry = pendingDelete { model.delete(entry) }
            pendingDelete = nil
        }
    }

    private func subtitle(_ entry: Wellbeing) -> String {
        var parts: [String] = []
        if let recovery = entry.recovery { parts.append(Str.t("wellbeing_sub_healing", recovery.label)) }
        if let bleeding = entry.bleeding { parts.append(Str.t("wellbeing_sub_bleeding", bleeding.label)) }
        if let breast = entry.breast { parts.append(Str.t("wellbeing_sub_breasts", breast.label)) }
        if let notes = entry.notes?.trimmingCharacters(in: .whitespacesAndNewlines), !notes.isEmpty {
            parts.append(notes)
        }
        return parts.joined(separator: Str.t("feeding_detail_separator"))
    }
}

/// Today's entry, written in place rather than behind a "+".
///
/// The other logs get a sheet, this one does not, and the difference is
/// deliberate: a parent logs a nappy twenty times a day and their own state
/// perhaps once, so the cost worth cutting here is the tap that reminds you the
/// screen exists — not the space the form takes up.
private struct WellbeingAddCard: View {
    let gaveBirth: Bool
    let breastfeeding: Bool
    let deliveryType: DeliveryType?
    let onAdd: (Wellbeing) -> Void

    @State private var mood = 3
    @State private var bleeding: Bleeding?
    @State private var recovery: Recovery?
    @State private var breast: BreastState?
    @State private var time = Clock.millis
    @State private var notes = ""

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.tight) {
            Text(Str.t("wellbeing_add_title"))
                .font(.headline)
                .foregroundStyle(SproutColor.onSurface)

            FieldLabel(Str.t("field_mood"))
            ChoiceChips(
                options: [1, 2, 3, 4, 5],
                selection: Binding(get: { mood }, set: { mood = $0 ?? mood }),
                label: { "\(moodEmoji($0)) \($0)" }
            )

            // Both questions are asked only of a parent who gave birth, and the
            // healing one is worded for the delivery — a caesarean question put
            // to someone who had a vaginal birth is the kind of small wrongness
            // this screen cannot afford.
            if gaveBirth {
                FieldLabel(healingFieldLabel(deliveryType))
                ChoiceChips(
                    options: Recovery.allCases,
                    selection: $recovery,
                    label: \.label,
                    allowsDeselection: true
                )

                FieldLabel(Str.t("field_bleeding"))
                ChoiceChips(
                    options: Bleeding.allCases,
                    selection: $bleeding,
                    label: \.label,
                    allowsDeselection: true
                )
            }

            if breastfeeding {
                FieldLabel(Str.t("field_breast_comfort"))
                ChoiceChips(
                    options: BreastState.allCases,
                    selection: $breast,
                    label: \.label,
                    allowsDeselection: true
                )
            }

            // Yesterday's rough night is worth recording today.
            DateTimeField(label: Str.t("field_date"), millis: $time)
            NotesField(text: $notes)

            HStack {
                Spacer()
                Button(Str.t("wellbeing_add"), action: save)
                    .buttonStyle(.borderedProminent)
            }
            .padding(.top, Spacing.hairline)
        }
        .padding(Spacing.regular)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(SproutColor.surface, in: RoundedRectangle(cornerRadius: Radius.card))
    }

    private func save() {
        onAdd(
            Wellbeing(
                time: time,
                mood: mood,
                // A question that was not asked is stored as unanswered, not as
                // whatever was left selected before the profile changed.
                bleeding: gaveBirth ? bleeding : nil,
                recovery: gaveBirth ? recovery : nil,
                breast: breastfeeding ? breast : nil,
                notes: notes.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
            )
        )
        mood = 3
        bleeding = nil
        recovery = nil
        breast = nil
        notes = ""
        time = Clock.millis
    }
}
